src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb

```diff
@@ -3802,18 +3825,23 @@ class C
     }
 }", finalText)
 
-                ' This should have happened as two text changes to the buffer.
                 Dim changes = snapshotBeforeCommit.Version.Changes
-                Assert.Equal(2, changes.Count)
-
-                Dim actualChanges = changes.ToArray()
-                Dim firstChange = actualChanges(0)
-                Assert.Equal(New Span(0, 0), firstChange.OldSpan)
-                Assert.Equal("using NewUsing;", firstChange.NewText)
-
-                Dim secondChange = actualChanges(1)
-                Assert.Equal(New Span(testDocument.CursorPosition.Value, 0), secondChange.OldSpan)
-                Assert.Equal("InsertedItem", secondChange.NewText)
+                Select Case completionImplementation
+                    Case CompletionImplementation.Legacy ' Specific of the legacy implementation
+                        ' This should have happened as two text changes to the buffer.
+                        Assert.Equal(2, changes.Count)
+
+                        Dim actualChanges = changes.ToArray()
+                        Dim firstChange = actualChanges(0)
+                        Assert.Equal(New Span(0, 0), firstChange.OldSpan)
+                        Assert.Equal("using NewUsing;", firstChange.NewText)
+
+                        Dim secondChange = actualChanges(1)
+                        Assert.Equal(New Span(testDocument.CursorPosition.Value, 0), secondChange.OldSpan)
+                        Assert.Equal("InsertedItem", secondChange.NewText)
+                    Case CompletionImplementation.Modern
+                        Assert.Equal(1, changes.Count)
```

Comment:
why is the modern and legacy doing different htings?  they should not be.

Reply:
The editor team uses different number of snapshots for the scenario.

---
In reply to: [222523183](https://github.com/dotnet/roslyn/pull/29016#discussion_r222523183) [](ancestors = 222523183)

Reply:
What are the implications of that?  does 'undo' work differently betwen legacy and modern?  Note that undo behavior is pretty critical and is also part of the core completion contract that we work hard to maintain.

Reply:
This is still under investigation

---
In reply to: [222850662](https://github.com/dotnet/roslyn/pull/29016#discussion_r222850662) [](ancestors = 222850662)

Reply:
❗️ This is missing assertions for the content of the change.

❗️ This is a bug, but not for Undo behavior. This issue will cause the **Track Changes** margin to show the entire file as changed when it should only be showing two lines as changed (separated by lines that did not change). It will also break certain optimizations in incremental parsing/analysis that depend on the changed span.

Reply:
Added assertions for the content of the change.

---
In reply to: [229263722](https://github.com/dotnet/roslyn/pull/29016#discussion_r229263722) [](ancestors = 229263722)

Reply:
Added [https://github.com/dotnet/roslyn/issues/30972](https://github.com/dotnet/roslyn/issues/30972)

---
In reply to: [230977947](https://github.com/dotnet/roslyn/pull/29016#discussion_r230977947) [](ancestors = 230977947,229263722)

Reply:
[https://github.com/dotnet/roslyn/issues/30972](https://github.com/dotnet/roslyn/issues/30972)

---
In reply to: [228694667](https://github.com/dotnet/roslyn/pull/29016#discussion_r228694667) [](ancestors = 228694667,222850662)